File: leetcode/39_combination_sum.cc
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>result;
        vector<int>currCandidates;
        dfs(0, target, candidates, currCandidates, result);
        return result;
    }

private:
    void dfs(int i, int target, vector<int>& candidates, vector<int>& currCandidates, vector<vector<int>>& result) {
        if (target == 0) {
            result.push_back(currCandidates);
            return;
        }
        if (target < 0 || i >= candidates.size()) {
            return;
        }
        currCandidates.push_back(candidates[i]);
        dfs(i, target - candidates[i], candidates, currCandidates, result);
        currCandidates.pop_back();
        dfs(i+1, target, candidates, currCandidates, result);
    }
};

File: leetcode/39_combination_sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        output = []
        subsets = []
        n = len(candidates)

        def backtrack(i, sub_target):
            if sub_target == 0:
                output.append(subsets.copy())
                return
            if sub_target < 0 or i >= n:
                return
            subsets.append(candidates[i])
            backtrack(i, sub_target - candidates[i])
            num = subsets.pop()
            backtrack(i + 1, sub_target + num)
        
        backtrack(0, target)
        return output

            
